pkg/kubelet/eviction/threshold_notifier_linux.go

```diff
@@ -67,51 +70,118 @@ func NewMemCGThresholdNotifier(path, attribute, threshold, description string, h
-		description: description,
-	}, nil
+	notifier := &linuxCgroupNotifier{
+		watchfd:   watchfd,
+		controlfd: controlfd,
+		eventfd:   eventfd,
+		epfd:      epfd,
+		stop:      make(chan struct{}),
+	}
+	return notifier, nil
 }
 
-func getThresholdEvents(eventfd int, eventCh chan<- struct{}, stop ThresholdStopCh) {
+func (n *linuxCgroupNotifier) Start(eventCh chan<- struct{}) {
+	err := unix.EpollCtl(n.epfd, unix.EPOLL_CTL_ADD, n.eventfd, &unix.EpollEvent{
+		Fd:     int32(n.eventfd),
+		Events: unix.EPOLLIN,
+	})
+	if err != nil {
+		glog.Warningf("eviction manager: error adding epoll eventfd: %v", err)
+		return
+	}
 	for {
-		buf := make([]byte, 8)
-		_, err := unix.Read(eventfd, buf)
+		event, err := wait(n.epfd, n.eventfd)
 		if err != nil {
+			glog.Warningf("eviction manager: error while waiting for memcg events: %v", err)
 			return
+		} else if event {
+			// Consume the event from the eventfd
+			buf := make([]byte, 8)
+			_, err := unix.Read(n.eventfd, buf)
+			if err != nil {
+				glog.Warningf("eviction manager: error reading memcg events: %v", err)
+				return
+			}
+			select {
+			case <-n.stop:
+				return
+			case eventCh <- struct{}{}:
+			}
 		}
 
-		select {
-		case eventCh <- struct{}{}:
-		case <-stop.Ch():
+		if _, ok := <-n.stop; !ok {
 			return
 		}
 	}
 }
 
-func (n *memcgThresholdNotifier) Start(stop ThresholdStopCh) {
-	eventCh := make(chan struct{})
-	go getThresholdEvents(n.eventfd, eventCh, stop)
-	for {
-		select {
-		case <-stop.Ch():
-			glog.V(3).Infof("eviction: stopping threshold notifier")
-			unix.Close(n.watchfd)
-			unix.Close(n.controlfd)
-			unix.Close(n.eventfd)
-			return
-		case <-eventCh:
-			glog.V(2).Infof("eviction: threshold crossed")
-			n.handler(n.description)
+// wait waits up to notifierRefreshInterval for an event on the Epoll FD for the
+// eventfd we are concerned about.  It returns an error if one occurrs, and true
+// if the consumer should read from the eventfd.
+func wait(epfd, eventfd int) (bool, error) {
+	events := make([]unix.EpollEvent, 7)
+	timeoutMS := int(notifierRefreshInterval / time.Millisecond)
+	n, err := unix.EpollWait(epfd, events, timeoutMS)
+	if n == -1 {
+		if err == unix.EINTR {
+			// Interrupt, ignore the error
+			return false, nil
 		}
+		return false, err
+	}
+	if n == 0 {
+		// Timeout
+		return false, nil
+	}
+	if n > 6 {
+		return false, fmt.Errorf("epoll_wait returned more events than we know what to do with")
+	}
+	for _, event := range events[:n] {
+		if event.Fd == int32(eventfd) {
+			if event.Events&unix.EPOLLHUP != 0 || event.Events&unix.EPOLLERR != 0 || event.Events&unix.EPOLLIN != 0 {
+				// EPOLLHUP: should not happen, but if it does, treat it as a wakeup.
+
+				// EPOLLERR: If an error is waiting on the file descriptor, we should pretend
+				// something is ready to read, and let unix.Read pick up the error.
+
+				// EPOLLIN: There is data to read.
+				return true, nil
+			}
+		}
+	}
+	// An event occurred that we don't care about.
```

Comment:
Oh, I see... But why not keep waiting in this case? Since that seems to be what the caller is doing anyways, just call wait() again...

Reply:
It gives us the opportunity to check the stop channel in the main loop.  I could pass the stop channel in, and have it do that itself, but we would still need to keep the loop in the `Start` function anyways to handle multiple events.

Reply:
I actually think passing this a channel is a slightly better API (also would get immediate notifications instead of depending on the timeout to poll "stop".)

But I can see how it might be hard to implement that, since you need to bridge Go channels and Linux epoll...

I guess the way to do that would be to open a pipe here and add the read side of the pipe to the epoll you're blocking on.

Then launch a goroutine that would block on the channel and, if it gets anything from it, write to the write end of the pipe (therefore leaving the epoll.)

But that's a lot of code, for little gain...

So I guess fine to keep this as is.

I'm not too pleased about the return of `(bool, error)`... Would there be a better way to return "timeout" as a specific error and check for that condition?

In that case, you might only want to return that on "timeout", in all other cases where you're currently returning false, you could just loop inside this function again...

But maybe that's too much bikeshedding and we should leave this alone... Up to you, whatever you prefer is fine with me.

Reply:
Ill change to a timeout error

Reply:
or err... There are a lot of cases where we just want to retry the wait... I think I am just going to leave it alone